File: lee/src/main/java/org/example/lee/题目/二分查找/b_探索二维矩阵.java
package org.example.lee.题目.二分查找;

public class b_探索二维矩阵 {

	public boolean searchMatrix(int[][] matrix, int target) {
		int m = matrix.length, n = matrix[0].length;
		int length = m * n;
		int left = 0, right = length - 1;

		while (left <= right) {
			int index = left + (right - left) / 2;
			int result = getNum(matrix, index);
			if (result == target) {
				return true;
			} else if (result > target) {
				right = index - 1;
			} else {
				left = index + 1;
			}
		}
		return false;
	}

	public int getNum(int[][] matrix, int index) {
		int m = matrix.length, n = matrix[0].length;
		int i = index / n;
		int j = index % n;
		return matrix[i][j];
	}
}

File: lee/src/main/java/org/example/lee/题目/二分查找/b_探索二维矩阵.go
package 二分查找

func searchMatrix(matrix [][]int, target int) bool {
	l1, l2 := len(matrix), len(matrix[0])
	left, right := 0, l1*l2
	for left < right {
		a := (left + right) / 2
		b := tool(matrix, a)
		if b == target {
			return true
		}
		if a == left || a == right {
			return false
		}
		if b < target {
			left = a + 1
		} else {
			right = a - 1
		}
	}
	return false
}
func tool(matrix [][]int, i int) int {
	x := i / len(matrix[0])
	y := i % len(matrix[0])
	return matrix[x][y]
}
